src/main/cljc/jarl/builtins/http.cljc

```diff
@@ -0,0 +1,199 @@
+    101 "101 Switching Protocols"
+    103 "103 Early Hints"
+    200 "200 OK"
+    201 "201 Created"
+    202 "202 Accepted"
+    203 "203 Non-Authoritative Information"
+    204 "204 No Content"
+    205 "205 Reset Content"
+    206 "206 Partial Content"
+    300 "300 Multiple Choices"
+    301 "301 Moved Permanently"
+    302 "302 Found"
+    304 "304 Not Modified"
+    307 "307 Temporary Redirect"
+    308 "308 Permanent Redirect"
+    400 "400 Bad Request"
+    402 "402 Payment Required"
+    404 "404 Not Found"
+    405 "405 Method Not Allowed"
+    406 "406 Not Acceptable"
+    409 "409 Conflict"
+    410 "410 Gone"
+    411 "411 Length Required"
+    412 "412 Precondition Failed"
+    413 "413 Payload Too Large"
+    414 "414 URI Too Long"
+    416 "416 Range Not Satisfiable"
+    417 "417 Expectation Failed"
+    425 "425 Too Early"
+    426 "426 Upgrade Required"
+    428 "428 Precondition Required"
+    429 "429 Too Many Requests"
+    431 "431 Request Header Fields Too Large"
+    451 "451 Unavailable For Legal Reasons"
+    500 "500 Internal Server Error"
+    501 "501 Not Implemented"
+    502 "502 Bad Gateway"
+    503 "503 Service Unavailable"
+    504 "504 Gateway Timeout"
+    505 "505 HTTP Version Not Supported"
+    506 "506 Variant Also Negotiates"
+    510 "510 Not Extended"
+    511 "511 Network Authentication Required"
+    (throw (ex-info "unknown http status code" {:code code}))))
+
+(def allowed-keys
+  #{"method"
+    "url"
+    "body"
+    "enable_redirect"
+    "force_json_decode"
+    "force_yaml_decode"
+    "headers"
+    "raw_body"
+    "tls_use_system_certs"
+    "tls_ca_cert"
+    "tls_ca_cert_file"
+    "tls_ca_cert_env_variable"
+    "tls_client_cert"
+    "tls_client_cert_file"
+    "tls_client_cert_env_variable"
+    "tls_client_key"
+    "tls_client_key_file"
+    "tls_client_key_env_variable"
+    "tls_insecure_skip_verify"
+    "tls_server_name"
+    "timeout"
+    "cache"
+    "force_cache"
+    "force_cache_duration_seconds"
+    "raise_error"
+    "caching_mode"})
+
+(def required-keys #{"method" "url"})
+
+(defn- format-set [s]
+  (str (str/join ", " (map #(str "\"" % "\"") s))))
+
+; TODO: verify correct type of each provided value
+(defn validate-request [request]
+  (let [keys (set (keys request))
+        invalid-keys (set/difference keys allowed-keys)
+        missing-keys (set/difference required-keys keys)
+        raise-error (get request "raise_error")]
+    (when (seq invalid-keys)
+      (throw (errors/type-ex "http.send: operand 1 invalid request parameters(s): {%s}", (format-set invalid-keys))))
+    (when (seq missing-keys)
+      (throw (errors/type-ex "http.send: operand 1 missing required request parameters(s): {%s}", (format-set missing-keys))))
+    (when (and (some? raise-error) (not (boolean? raise-error)))
+      (throw (errors/type-ex "http.send: invalid value for raise_error field")))))
+
+(defn verify-url [s]
+  (if (nil? (re-find #"https?://(\S+)" s))
```

Comment:
Is this really necessary? There is the question of parity with OPA, of course, but isn't it enough that the HTTP client couldn't establish a connection to whatever URL was handed down?

Reply:
Probably, I mainly wanted something here that was consistent regardless of implementation.